docs/langref/relay/type_system.rst

```diff
@@ -0,0 +1,187 @@
+===========
+Type System
+===========
+
+We have briefly introduced types while detailing the the expression language
+of Relay, but have fully laid out the type system.
+
+Although the majority of Relay programs are written without type annotations,
+Relay is statically typed.
+
+Static types are useful because they enable efficient layout, memory reuse, and
+code generation. They aid in debugging program transformations, but can also
+give us the expressivity afforded by more dynamic langauges.
+
+We are able to omit these type annotations by a process known as type inference.
+Type inference is a technique that has its roots in the programming language
+community, and can be viewed as a method for generalizing shape inference to
+run over arbitrary user programs containing control flow and recursion.
+
+Static types are useful when performing compiler optimization because they
+communicate properties about the data we manipulate, such as runtime shape,
+data layout, storage without needing to run the program.
+
+Most current IRs use "shape inference" to recover Tensor dimensions from the user
+provided program. Machine learning users have enjoyed shape inference for
+tensors because it allows them to generate performant code without giving up
+on the expressivity of the input language.
+
+Because Relay is intended as an IR we require *some* type information to provide
+full inference. We don't believe this to be an issue as many of the IR builder
+inferfaces require some type information, or can generate IR based on their own
+higher level inferences.
+
+We view this limited shape inference as a simpler form of type
+inference. Instead of relying on an ad-hoc procedure for recovering type
+information from a potentially dynamic program, we apply ideas from compiler and IR design.
+
+Below we briefly dicsuss the different kinds of types in Relay.
+
+=====
+Types
+=====
+
+Relay's type system has a "language of types" which allow us to write down the type of
+a Relay program. Below we detail the langauge of types and how we assign them to Relay
```

Comment:
"them" is ambiguous in this sentence, would be clearer to explicitly write "how we assign types"